File: svix-cli/src/cmds/api/streaming_events.rs
// this file is @generated
use clap::{
    Args,
    Subcommand,
};
use svix::api::*;

#[derive(Args, Clone)]
pub struct StreamingEventsCreateOptions {
    #[arg(long)]
    pub idempotency_key: Option<String>,
}

impl From<StreamingEventsCreateOptions> for svix::api::StreamingEventsCreateOptions {
    fn from(value: StreamingEventsCreateOptions) -> Self {
        let StreamingEventsCreateOptions {
            idempotency_key,
        } = value;
        Self {
            idempotency_key,
        }
    }
}

#[derive(Args, Clone)]
pub struct StreamingEventsGetOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    #[arg(long)]
    pub after: Option<chrono::DateTime<chrono::Utc>>,
}

impl From<StreamingEventsGetOptions> for svix::api::StreamingEventsGetOptions {
    fn from(value: StreamingEventsGetOptions) -> Self {
        let StreamingEventsGetOptions {
            limit,
            iterator,
            after,
        } = value;
        Self {
            limit,
            iterator,
            after: after.map(|dt| dt.to_rfc3339()),
        }
    }
}

#[derive(Args)]
#[command(args_conflicts_with_subcommands = true, flatten_help = true)]
pub struct StreamingEventsArgs {
    #[command(subcommand)]
    pub command: StreamingEventsCommands,
}

#[derive(Subcommand)]
pub enum StreamingEventsCommands {
    /// Creates events on the Stream.
    Create {
        stream_id: String,
        create_stream_events_in: crate::json::JsonOf<CreateStreamEventsIn>,
        #[clap(flatten)]
        options: StreamingEventsCreateOptions,
    },
    /// Iterate over a stream of events.
    ///
    /// The sink must be of type `poller` to use the poller endpoint.
    Get {
        stream_id: String,
        sink_id: String,
        #[clap(flatten)]
        options: StreamingEventsGetOptions,
    },
}

impl StreamingEventsCommands {
    pub async fn exec(
        self,
        client: &Svix,
        color_mode: colored_json::ColorMode,
    ) -> anyhow::Result<()> {
        match self {
            Self::Create {
                stream_id,
                create_stream_events_in,
                options,
            } => {
                let resp = client
                    .streaming()
                    .events()
                    .create(
                        stream_id,
                        create_stream_events_in.into_inner(),
                        Some(options.into()),
                    )
                    .await?;
                crate::json::print_json_output(
                    &resp, color_mode,
                )?;
            }
            Self::Get {
                stream_id,
                sink_id,
                options,
            } => {
                let resp = client
                    .streaming()
                    .events()
                    .get(
                        stream_id,
                        sink_id,
                        Some(options.into()),
                    )
                    .await?;
                crate::json::print_json_output(
                    &resp, color_mode,
                )?;
            }
        }

        Ok(())
    }
}

File: python/svix/api/streaming_events.py
# This file is @generated
import typing as t
from dataclasses import dataclass
from datetime import datetime

from ..models import CreateStreamEventsIn, CreateStreamEventsOut, EventStreamOut
from .common import ApiBase, BaseOptions, serialize_params


@dataclass
class StreamingEventsCreateOptions(BaseOptions):
    idempotency_key: t.Optional[str] = None

    def _header_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "idempotency-key": self.idempotency_key,
            }
        )


@dataclass
class StreamingEventsGetOptions(BaseOptions):
    limit: t.Optional[int] = None
    """Limit the number of returned items"""
    iterator: t.Optional[str] = None
    """The iterator returned from a prior invocation"""
    after: t.Optional[datetime] = None

    def _query_params(self) -> t.Dict[str, str]:
        return serialize_params(
            {
                "limit": self.limit,
                "iterator": self.iterator,
                "after": self.after,
            }
        )


class StreamingEventsAsync(ApiBase):
    async def create(
        self,
        stream_id: str,
        create_stream_events_in: CreateStreamEventsIn,
        options: StreamingEventsCreateOptions = StreamingEventsCreateOptions(),
    ) -> CreateStreamEventsOut:
        """Creates events on the Stream."""
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/stream/{stream_id}/events",
            path_params={
                "stream_id": stream_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=create_stream_events_in.model_dump_json(
                exclude_unset=True, by_alias=True
            ),
        )
        return CreateStreamEventsOut.model_validate(response.json())

    async def get(
        self,
        stream_id: str,
        sink_id: str,
        options: StreamingEventsGetOptions = StreamingEventsGetOptions(),
    ) -> EventStreamOut:
        """Iterate over a stream of events.

        The sink must be of type `poller` to use the poller endpoint."""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/stream/{stream_id}/sink/{sink_id}/events",
            path_params={
                "stream_id": stream_id,
                "sink_id": sink_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
        )
        return EventStreamOut.model_validate(response.json())


class StreamingEvents(ApiBase):
    def create(
        self,
        stream_id: str,
        create_stream_events_in: CreateStreamEventsIn,
        options: StreamingEventsCreateOptions = StreamingEventsCreateOptions(),
    ) -> CreateStreamEventsOut:
        """Creates events on the Stream."""
        response = self._request_sync(
            method="post",
            path="/api/v1/stream/{stream_id}/events",
            path_params={
                "stream_id": stream_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
            json_body=create_stream_events_in.model_dump_json(
                exclude_unset=True, by_alias=True
            ),
        )
        return CreateStreamEventsOut.model_validate(response.json())

    def get(
        self,
        stream_id: str,
        sink_id: str,
        options: StreamingEventsGetOptions = StreamingEventsGetOptions(),
    ) -> EventStreamOut:
        """Iterate over a stream of events.

        The sink must be of type `poller` to use the poller endpoint."""
        response = self._request_sync(
            method="get",
            path="/api/v1/stream/{stream_id}/sink/{sink_id}/events",
            path_params={
                "stream_id": stream_id,
                "sink_id": sink_id,
            },
            query_params=options._query_params(),
            header_params=options._header_params(),
        )
        return EventStreamOut.model_validate(response.json())
